spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java

```diff
@@ -0,0 +1,146 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+package edu.umd.cs.findbugs.detect;
+
+import org.apache.bcel.Const;
+
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.StringAnnotation;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * This detector can find ...
+ */
+public class DateFormatStringChecker extends OpcodeStackDetector {
+    private final BugReporter bugReporter;
+
+    /**
+     * Stores required and forbidden flags. Then confirms the absence of a forbidden
+     * flag and presence of all required flags via verify method
+     */
+    private static final class DateFormatRule {
+
+        /**
+         * forbiddenFlag must be missing to match given DateFormatRule. If this field
+         * is {@code null}, rule does not have a forbiddenFlag and the associated check
+         * will be bypassed.
+         */
+        @Nullable
+        String forbiddenFlag;
+
+        /**
+         * List of flags required in a pattern to match given DateFormatRule.
+         */
+        @NonNull
+        List<String> requiredFlags;
+
+        DateFormatRule(String forbiddenFlag, String... requiredFlags) {
+            this.forbiddenFlag = forbiddenFlag;
+            this.requiredFlags = Arrays.asList(requiredFlags);
+        }
+
+        /**
+         * Checks if string does not contain a forbiddenFlag and has all requiredFlags
+         *
+         * @param dateFormatString - string to be checked
+         * @return {@code true} if code equals to one of the elements
+         */
+        boolean verify(String dateFormatString) {
+            if (this.forbiddenFlag != null && dateFormatString.contains(this.forbiddenFlag)) {
+                return false;
+            }
+            for (String flag : requiredFlags) {
+                if (!dateFormatString.contains(flag)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    public DateFormatStringChecker(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
+    /**
+     * Checks whether the current element is a proper function (one of three specified)
+     * that has at least 1 parameter. Such methods would have 1st parameter to appear as
+     * dateFormat string, which will be tested for forbidden combinations.
+     */
```

Comment:
Are you sure this javadoc describes well what the function does? Please, either rewrite it or remove it.